[MINOR][DOCS] Add a clarifying note to str_to_map documentation

MichaelChirico opened the pull request:
I was quite surprised by the following behavior:

`SELECT str_to_map('1:2|3:4', '|')`

vs

`SELECT str_to_map(replace('1:2|3:4', '|', ','))`

The documentation does not make clear at all what's going on here, but a [dive into the source code shows](https://github.com/apache/spark/blob/fa0d4bf69929c5acd676d602e758a969713d19d8/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala#L461-L466) that `split` is being used and in turn the interpretation of `split`'s arguments as RegEx is clearly documented.

## What changes were proposed in this pull request?

Documentation clarification

## How was this patch tested?

N/A

MichaelChirico committed:
Add a clarifying note to str_to_map documentation

I was quite surprised by the following behavior:

`SELECT str_to_map('1:2|3:4', '|')`

vs

`SELECT str_to_map(replace('1:2|3:4', '|', ','))`

The documentation does not make clear at all what's going on here, but a [dive into the source code shows](https://github.com/apache/spark/blob/fa0d4bf69929c5acd676d602e758a969713d19d8/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala#L461-L466) that `split` is being used and in turn the interpretation of `split`'s arguments as RegEx is clearly documented.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented:
ok to test

HyukjinKwon commented:
@MichaelChirico, mind if I ask to double check `str_to_map` in other DBMS? I think it should be best to match with them

MichaelChirico commented:
@HyukjinKwon you mean if `pairDelim` and `keyValueDelim` are treated as regex in other DBMS, i.e. if Spark behavior is consistent?

Since this edit is just for documentation, best to just add it for now since it indeed reflects the actual behavior of Spark as of now.

I'll look into this and file a separate JIRA if there's inconsistency across DBMS

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8318/
Test PASSed.

SparkQA commented:
**[Test build #102813 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102813/testReport)** for PR 23888 at commit [`053e59d`](https://github.com/apache/spark/commit/053e59dcc5883fb6f2997aefa7a19f9a623e5182).

MichaelChirico commented:
Should have searched JIRA first beforehand; this PR is related to  `HIVE-13482`:

https://issues.apache.org/jira/browse/HIVE-13482

MichaelChirico commented:
I'm also seeing a few references here, should update this wiki page as well?

https://cwiki.apache.org/confluence/display/Hive/LanguageManual+UDF#LanguageManualUDF-StringFunctions

MichaelChirico commented:
Other references I've seen about the values as regex on SO:

https://stackoverflow.com/a/49790108/3576984

https://stackoverflow.com/a/54143796/3576984

HyukjinKwon commented:
In Spark's case, the function documentation is automatically generated based on the given description at https://spark.apache.org/docs/2.4.0/api/sql/ . So no need to manually update.

The problem is, as soon as we document, that's an official behaviour. I was wondering if this is something we intend.

Do you know if this behaviour is intended in Hive side as well? Then I think we can go ahead with this.

MichaelChirico commented:
Right, makes sense. The stuff I cited above (cwiki and both SO answers) are under Hive specifically, no mention of Spark.

As to whether the behavior is _intended_ in Hive, it's unclear. But certainly the [implementation in Hive](https://github.com/apache/hive/blob/f37c5de6c32b9395d1b34fa3c02ed06d1bfbf6eb/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFStringToMap.java#L103-L106) does the same thing (uses `split` and therefore is regex):

```
    String[] keyValuePairs = text.split(delimiter1);

    for (String keyValuePair : keyValuePairs) {
      String[] keyValue = keyValuePair.split(delimiter2, 2);
```

Further, here [1](https://github.com/apache/hive/blob/672755d6940923483f8d5a08959e182580edc72c/ql/src/test/results/clientpositive/str_to_map.q.out), [2](https://github.com/apache/hive/blob/f37c5de6c32b9395d1b34fa3c02ed06d1bfbf6eb/ql/src/test/queries/clientpositive/vector_udf1.q), [3](https://github.com/apache/hive/blob/f37c5de6c32b9395d1b34fa3c02ed06d1bfbf6eb/ql/src/test/queries/clientpositive/char_udf1.q), [4](https://github.com/apache/hive/blob/f37c5de6c32b9395d1b34fa3c02ed06d1bfbf6eb/ql/src/test/queries/clientpositive/varchar_udf1.q), [5](https://github.com/apache/hive/blob/83e53972c07df8b7d9a01ad14dda5cb550406e87/ql/src/test/results/clientpositive/llap/varchar_udf1.q.out), [6](https://github.com/apache/hive/blob/f18842e7379060bb2504b0dbae4f4280abba883b/ql/src/test/results/clientpositive/char_udf1.q.out), [7](https://github.com/apache/hive/blob/8b968c7e46929c3af86da46e316faeb8d17f03df/ql/src/test/results/clientpositive/llap/vector_udf1.q.out), [8](https://github.com/apache/hive/blob/f37c5de6c32b9395d1b34fa3c02ed06d1bfbf6eb/ql/src/test/queries/clientpositive/annotate_stats_select.q), [9](https://github.com/apache/hive/blob/f37c5de6c32b9395d1b34fa3c02ed06d1bfbf6eb/ql/src/test/queries/clientpositive/vector_annotate_stats_select.q), [10](https://github.com/apache/hive/blob/672755d6940923483f8d5a08959e182580edc72c/ql/src/test/results/clientpositive/annotate_stats_select.q.out), [11](https://github.com/apache/hive/blob/672755d6940923483f8d5a08959e182580edc72c/ql/src/test/results/clientpositive/llap/vector_annotate_stats_select.q.out) are some tests of `str_to_map` but none of the tests use regex-ambiguous `delimiter{1,2}` values...

My conclusion is that it's eminently ambiguous whether the _intended_ behavior in either Hive or SparkSQL is to treat the delimiters as regular expressions.

BUT the behavior has been around for [8 years](https://github.com/apache/hive/commit/4f8294e578db449294a1186f0ac4efb041445dcb) and at least going off of the SO answers, it seems to be accepted as "known" behavior so things will probably break if we change it.

I'll follow up this PR with one over at Hive to make sure things are in sync.

SparkQA commented:
**[Test build #102813 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102813/testReport)** for PR 23888 at commit [`053e59d`](https://github.com/apache/spark/commit/053e59dcc5883fb6f2997aefa7a19f9a623e5182).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102813/
Test FAILed.

srowen reviewed (approved):
Late to the party here but I think it's worth documenting. Users do need to know what the behavior is, and even if nobody officially promised it, this is de facto the API and a change to it would be breaking anyway.

SparkQA commented:
**[Test build #4792 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4792/testReport)** for PR 23888 at commit [`053e59d`](https://github.com/apache/spark/commit/053e59dcc5883fb6f2997aefa7a19f9a623e5182).

SparkQA commented:
**[Test build #4792 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4792/testReport)** for PR 23888 at commit [`053e59d`](https://github.com/apache/spark/commit/053e59dcc5883fb6f2997aefa7a19f9a623e5182).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

HyukjinKwon reviewed (approved):
makes sense. Looks good to me too

HyukjinKwon commented:
Merged to master.
